File: ObjectC/QiSpaceObjectiveC/Qeep_Demo.m
//
//  Qeep_Demo.m
//  QiSpace_ObjectiveC_Demo
//
//  Created by Roy Zheng on 2022-06-13.
//

#import "Qeep_Demo.h"
#import "qeepObjCWrapper.h"
#import "Sequr_Demo.h"
#import "ViewController.h"
#import "sequrObjCWrapper.h"
#include "qispace_info.h"

/**
 * This page demonstrates a use of the QEEP API to encrypt and dedcrypt a message with a QiSpace retrieved Quantum Key (QK) and iv.
 * Also, the SEQUR API is also demonstrated as a by-product of decoding the retrieved Quantum Key (QK) and iv.
 *
 * Prerequisites:
 *  - An enterprise device token which will be used for QK retrieval from QiSpace Enterprise REST API
 *  - A subkey intended for SEQUR decoding the QK after its retrieval. Realistically, subkey would be initialized on device setup
 *
 *  **Ideally, setting up the subkey from Sequr_Demo must be done before getting QK as QK decryption requires subkey to be initialized
 *
 *  Logical Flow:
 *   1. Press 'Get QK' to create and unpack a Quantum Key from QiSpace Enterprise
 *   2. Enter a message into the text field
 *   3. Press 'Encrypt' to QEEP encrypt the text field message
 *   4. Press 'Decrypt' to QEEP decrypt the encrypted text field message
 */
// Note: Assume subkey is given for this demo page, so it must be retrieved with on the SEQUR page. Realistically, it would be initialized on device setup



// Setup macros for constants
#define QK_KEY_LENGTH 64
#define QK_BUFFER_SIZE (QK_KEY_LENGTH+32)
#define MESSAGE_IV_LENGTH 16

// Helper function to convert string to byte array
void string2ByteArray(char* input, uint8_t* output)
{
    int loop = 0;
    int i = 0;
    while(input[loop] != '\0'){
        output[i++] = input[loop++];
    }
}



@interface Qeep_Demo ()
@end

@implementation Qeep_Demo



uint8_t qk[QK_BUFFER_SIZE];
size_t qk_len;
size_t iv_len;
bool qk_retrieved;


// Currently, max message length is 64
uint8_t *encrypted_message;
uint8_t *decrypted_message;

uint8_t *message_iv;        // Note: Must be 16 bytes
NSInteger message_field_len;

@synthesize message_field, encrypted_message_text, decrypted_message_text;



- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Initial setup
    qk_retrieved = false;
    encrypted_message = NULL;
    decrypted_message = NULL;
    
    message_iv = (uint8_t *)calloc(MESSAGE_IV_LENGTH, 1);      // Note: The message iv must be 16 bytes
}



/**
 * Performs base64 decoding, SEQUR decode and stores the QK and IV from HTTP request payload
 *
 *@param data the json data returned from QiSpace Enterprise after issuing a POST request for QK
 *@param response the default HTTP response regarding the QK request
 */
void unpack_qk(NSData * _Nullable data, NSURLResponse * _Nullable response){
    NSHTTPURLResponse *qk_response = (NSHTTPURLResponse *) response;
    NSLog(@"The response is: %@", qk_response);

    // Validate QiSpace API payload
    if((long)[qk_response statusCode] != 200){
        NSLog(@"Error with URL RESPONSE");
        return;
    }
    
    // Process qk json data as dictionary and extract QK and its IV
    NSDictionary *qk_payload = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:nil];
    NSLog(@"object: %@ \n", qk_payload);
    NSString *encrypted_qk_string = [qk_payload objectForKey: @"payload"];
    NSString *encrypted_qk_iv_string = [qk_payload objectForKey: @"iv"];
    
    // Convert to C-string because Base64 conversion is C-based
    const char *encrypted_qk_c_string = [encrypted_qk_string UTF8String];
    const char *encrypted_qk_iv_c_string = [encrypted_qk_iv_string UTF8String];
    
    // Converting QK and iv from Base64 to byte array
    uint8_t *unpacked_qk_byte_array = base64_decode_to_bytes(encrypted_qk_c_string, strlen(encrypted_qk_c_string), &qk_len);
    uint8_t *iv_byte_array = base64_decode_to_bytes(encrypted_qk_iv_c_string, strlen(encrypted_qk_iv_c_string), &iv_len);
   
    // SEQUR decode the retrieved QK using subkey
    memset(qk, 0, QK_BUFFER_SIZE);
    int ret = SEQUR_decode(subkey, subkey_len, iv_byte_array, unpacked_qk_byte_array, qk_len, qk);
    
    qk_retrieved = true;
}

/**
 * Triggers retrieval of QK from the QiSpace Enterprise REST API
 *
 * @param token the enterprise device bearer token used as authentication when accessing QiSpace API
 */
void get_qk(NSString *token){
    
    NSString *base = [NSString stringWithUTF8String:QISPACE_ENTERPRISE_BASE_URL];
    NSString *qk_endpoint = [NSString stringWithFormat:@"%@/api/v1/qk/", base];
    
    /**
     *       QISpace API response consists of
     *       a payload, a pre-encrypted Qeep Key (QK),
     *       and an iv, initialization vector to reduce encryption repetition,
     *       all in Base64 format
     */
    // Configuring QiSpace Quantum Key (QK) REST API request with specifications
    NSMutableURLRequest *qk_request = [[NSMutableURLRequest alloc] initWithURL: [NSURL URLWithString: qk_endpoint]];
    [qk_request setHTTPMethod: @"POST"];
    [qk_request setValue:@"application/json"
          forHTTPHeaderField:@"Content-Type"];
    [qk_request setValue:token
          forHTTPHeaderField:@"Authorization"];
    
    NSDictionary *qk_param_dict = @{@"key_length":@QK_KEY_LENGTH, @"label":@"demo-device"};
    NSData *qk_param = [NSJSONSerialization dataWithJSONObject:qk_param_dict options:kNilOptions error:nil];
    [qk_request setHTTPBody:qk_param];
    
    NSURLSession *qk_session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]
                                                          delegate:nil
                                                     delegateQueue:[NSOperationQueue mainQueue]];
    
    NSURLSessionDataTask *qk_task = [qk_session dataTaskWithRequest:qk_request
                            completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        unpack_qk(data, response);
    }];
    // Trigger QiSpace QK request API call
    [qk_task resume];
}




/**
 * Handles 'Get QK" button press event
 * 
 * Prerequisite: requires sub_key to be setup so that retrieved QK can be decoded
 */
- (IBAction)event_qk_pressed:(id)sender {
    // Assert subkey is initialized before continuing
    if(subkey == NULL){
        encrypted_message_text.text = @"Need to initialize subkey before getting QK";
        decrypted_message_text.text = @"Need to initialize subkey before getting QK";
        return;
    }
    get_qk(token);
}

/**
 * Generates an IV and encrypts the message using QEEP API with the retrieved QK
 *
 * @param qk the Quantum Key for symmetric encryption used in QEEP API
 * @param qk_len length value of the Quantum Key (QK)
 * @param message_iv byte array  where the message's initialization vector will be stored
 * @param message_field_len length value of the text field message
 * @param encrypted_message the variable designed to store the QEEP encrypted message
 * @param message_field_string the string to be QEEP encrypted
 * @returns the encrypted message as a UTF-8 string  or an error statement
 */
NSString* encrypt_operation(uint8_t *qk, size_t qk_len, uint8_t *message_iv, NSInteger message_field_len, uint8_t *encrypted_message, NSString *message_field_string){
    // Generate random 16 bytes (MANDATORY) iv per encrypt operation or per message
    // The message iv is simply fixed for this demo
    char *message_iv_as_hex = "0102030405060708090a010203040506";
    
    // Convert message iv from hex to byte array
    convert_hex_to_bytes(message_iv, strlen(message_iv_as_hex), message_iv_as_hex);
    
    // Dynamically allocate memory for encrypted
    encrypted_message = (uint8_t*)calloc(message_field_len, sizeof(uint8_t));
    
    // Convert the message to raw byte array
    uint8_t *raw_message = (uint8_t*)calloc(message_field_len, 1);
    string2ByteArray([message_field_string UTF8String], raw_message);
    
    // QEEP encrypt the text field message with the retrieved QK and generated IV
    int ret = QEEP_encrypt(qk, qk_len, message_iv, raw_message, (int32_t)message_field_len, encrypted_message);
    
    // Handle UI display with error codes
    NSString *output;
    if(ret == 1){
        // Formatting byte array into hex string
        char *outstr = convert_bytes_to_hex(encrypted_message, (int)message_field_len);
        output = [NSString stringWithUTF8String: outstr];
        free(outstr);
    }
    else{
        output = [NSString stringWithFormat: @"Error code: %d occurred in encryption. Consult error.h for more information", ret];
    }
    
    // Perform cleanup by freeing allocated memory
    free(raw_message);
    free(encrypted_message);
    
    return output;
}


/**
 * Handles 'Encrypt' button press event
 * Prerequisite: QK must exist as it is used as the symmetric encryption key
 */
- (IBAction)event_encrypt_pressed:(id)sender {
    // Check qk initialized for QEEP encryption/decryption requires QK
    if(!qk_retrieved){
        encrypted_message_text.text = @"Need to retrieve qk before encrypting";
        return;
    }
    
    // Get message from input field
    NSString *message_field_string = message_field.text;
    message_field_len = [message_field_string length];
    
    // Encrypt the message with QK and message_iv
    encrypted_message_text.text = encrypt_operation(qk, qk_len, message_iv, message_field_len, encrypted_message, message_field_string);
}

/**
 * Applies QEEP decryption on the encrypted message
 *
 * @param qk the Quantum Key used for symmetric cryptography with QEEP
 * @param qk_len the length value of the Quantum Key
 * @param message_iv the same initialization vector used to encrypt the message
 * @param encrypted_message_string the message to be QEEP decrypted in string form
 * @param message_field_len the length of the original message/decrypted message
 * @param decrypted_message the variable to store the QEEP decrypted message
 * @returns the decrypted message as utf-8 string or an error statement
 */
NSString* decrypt_operation(uint8_t *qk, size_t qk_len, uint8_t *message_iv, NSString *encrypted_message_string, NSInteger message_field_len, uint8_t *decrypted_message){
    decrypted_message = (uint8_t*)calloc(message_field_len, sizeof(uint8_t));
    
    // Convert char* to byte array
    const char *text = [encrypted_message_string UTF8String];
    uint8_t *encr_msg = (uint8_t *)calloc(message_field_len, sizeof(uint8_t));
    convert_hex_to_bytes(encr_msg, strlen(text), text);
    
    printf("encr_msg: %s\n", encr_msg);
    
    // Use Qeep API's QEEP_decrypt to decrypt the encrypted message
    // Message iv should be identical to one used during encryption
    int ret = QEEP_decrypt(qk, qk_len, message_iv, encr_msg, message_field_len, decrypted_message);
    
    // Handle UI display depending on error codes
    NSString *output;
    if(ret == 1){
        // Formatting the byte array into a string for printing
        char outstr[message_field_len];
        memcpy(outstr, decrypted_message, message_field_len);
        outstr[message_field_len] = '\0';
        output = [NSString stringWithUTF8String: outstr];
    }
    else{
        output = [NSString stringWithFormat: @"Error code: %d occurred in encryption. Consult error.h for more information", ret];
    }
    
    // Clean up by freeing memory
    free(encr_msg);
    free(decrypted_message);
    
    return output;
}


/**
 * Handles 'Decrypt' button press event
 * Prerequisite: an encrypted message must be present such that there is something to decrypt (also implies QK must be retrieved)    
 */
- (IBAction)event_decrypt_pressed:(id)sender {
    if(!qk_retrieved){
        decrypted_message_text.text = @"Need to retrieve QK before decrypting";
        return;
    }
    
    // Ensure message has been encrypted before attempting decryption
    if([encrypted_message_text.text isEqualToString: @"**encrypted message**"]){
        decrypted_message_text.text = @"Need an encrypted message to decrypt";
        return;
    }
    
    decrypted_message_text.text = decrypt_operation(qk, qk_len, message_iv, encrypted_message_text.text, message_field_len, decrypted_message);
}
// Handles cleanup activities as View exits
- (void)viewWillDisappear:(BOOL)animated{
    qk_retrieved = false;
    NSLog(@"freeing message_iv");
    free(message_iv);
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: c/Swift/QiSpaceiOSDemoApp/QeepDemo.swift
//
//  QeepDemo.swift
//  QEEPiOSDemoApp
//
//  Created by Roy Zheng on 2022-06-01.
//

import SwiftUI

/// This page demonstrates a use of the QEEP API to encrypt and decrypt a message with a  QiSpace retrieved Quantum Key (QK) and iv
/// Additionally, it shows using SEQUR API as a means to SEQUR decode the retrieved and encrypted Quantum Key.
///
///** MUST setup the subscriber key (subkey) from SequrDemo before using QK since QK decryption requires subkey to be initialized**
///
/// Prerequisites:
///  - An enterprise device token which will be used as authentication for QK retrieval from QiSpace REST API
///  - A subkey intended for SEQUR decoding the QK after its retrieval
///
/// Logical flow:
///  1. Press 'Get QK' to obtain QK for encrypting/decrypting purposes
///  2. Enter a message into the text field
///  3. Press 'Encrypt' to QEEP encrypt the text field message
///  4. Press 'Decrypt' to QEEP decrypt the previously encrypted message

struct QeepDemo: View {
    
    let qeep_demo_BG_colour = Color(red: 0, green: 0, blue: 0, opacity: 77.5/100)
    
    // Attributes for modifiable UI texts
    @State var user_message: String = "Enter message"
    @State var encrypted_message_string : String = "encrypted message"
    @State var decrypted_message_string : String = "decrypted message"
    
    // Attributes for API-related operations
    @State var encrypted_message_bytes: [UInt8]?
    @State var decrypted_message_bytes: [UInt8]?
    @State var message_iv: [UInt8]? = nil
    let qk_endpoint = qispace_info.base_url + "/api/v1/qk/"
    @Binding var sub_key : [UInt8]
    @State var qk : [UInt8]? = nil
    
    // Note: Token is set as this for this demo. Normally, token is generated upon device setup
    // Actual device token was removed for privacy reasons
    let token = "Bearer " + qispace_info.access_token
    
    var sequr_wrapper : sequrSwiftWrapper? = sequrSwiftWrapper()
    var qeep_wrapper : qeepSwiftWrapper? = qeepSwiftWrapper()
    
    /// Stores the retrieved Quantum Key (QK) from QiSpace Enterprise into subkey global variable
    ///
    /// - Prerequisites:
    ///  - Valid device token: a valid enterprise device token string needed to retrieve QK from QiSpace Enterprise
    ///  - subkey: a byte array used to retrieve QK from QiSpace Enterprise (need to setup subkey in SEQUR before getting QK)
    /// - Parameters: None
    /// - Returns: None
    func get_qk(){
        let sequr_wrapper = sequrSwiftWrapper()
        
        // Sets up HTTP request objects with specifications
        var qk_request = URLRequest(url: URL(string: qk_endpoint)!)
        qk_request.httpMethod = "POST"
        let json : [String: Any] = ["key_length" : 32, "label": "demo-device", "isExternal" : true]     // Note: In this demo, key length is set to be 32 bytes
        let bodyJSON = try? JSONSerialization.data(withJSONObject: json)
        qk_request.httpBody = bodyJSON
        qk_request.setValue("application/json", forHTTPHeaderField: "Content-Type")     // Note: Content-Type must be of application/json
        qk_request.setValue(token, forHTTPHeaderField: "Authorization")
        

        /// QiSpace API response consists of
        /// a payload, a pre-encrypted qeep key (QK),
        /// and an iv, a variable that diminishes data enryption repetition,
        /// all in Base64 format
        let qk_operation = URLSession.shared.dataTask(with: qk_request, completionHandler: { data, response, error -> Void in
            
            // Check for API request error and display
            if let error = error{
                print("Request error: ", error)
                return
            }
            
            guard let response = response as? HTTPURLResponse else { return }
            if response.statusCode == 200{      // Note: Currently, only 200 is thrown on success. Might need to configure checks for different success codes in the future
                DispatchQueue.main.async{
                    do{
                        let qk_payload = try JSONSerialization.jsonObject(with: data!) as? [String: Any]
                        
                        // Convert the payload and iv from Base64 to byte array
                        var encrypted_QK : [UInt8] = SequrDemo.base64_to_bytes(
                            string: String(describing: qk_payload!["payload"]!)
                        )
                        var qk_iv : [UInt8] = SequrDemo.base64_to_bytes(
                            string: String(describing: qk_payload!["iv"]!)
                        )
                        
                        // SEQUR decode the payload to get  Qeep Key (QK) with sub_key
                        self.qk = [UInt8].init(repeating: 0, count: encrypted_QK.count)
                        var ret = sequr_wrapper.decode(
                            key: &self.sub_key,
                            iv: &qk_iv,
                            encoded_message: &encrypted_QK,
                            decoded_message: &self.qk!
                        )
                        
                    }catch{
                        print("error with qe/iv")
                    }
                }
            }
        })
        
        // Trigger QK retrieval from QiSpace REST API
        qk_operation.resume()
    }
    
    
    /// 'Encrypt' button press event
    func encrypt_btn_pressed(){
        var message_bytes: [UInt8] = Array(user_message.utf8)
        encrypted_message_string = encrypt_operation(key: &(qk), iv: &message_iv, message: &message_bytes, encrypted_message: &encrypted_message_bytes)
    }
    
    
    /// Performs encryption using the QEEP algorithm on a passed-in byte array message
    ///
    /// - Prerequisites:
    ///  - QK: Must be retrieved before attempting encryption/decryption
    ///  - iv:  must be setup in order to use during encryption process
    /// - Parameters:
    ///  - key: a cryptographic key used to perform the encryption
    ///  - iv: initialization vector to improve randomness
    ///  - message: the message to be encrypted as a byte array
    ///  - encrypted_message: the message after encryption
    /// - Returns: A QEEP encrypted form of the message or an error statement
    func encrypt_operation(key qk: inout [UInt8]?, iv message_iv: inout [UInt8]?, message message_bytes:inout [UInt8], encrypted_message encrypted: inout [UInt8]?) -> String{
        
        // Verify that QK is present before attempting to encrypt/decrypt
        if(qk == nil){
            return "Need to retrieve QK before encrypting"
        }
        
        // Initialize data
        encrypted = [UInt8].init(repeating: 0, count: message_bytes.count)
        decrypted_message_bytes = [UInt8].init(repeating: 0, count: message_bytes.count)
        message_iv = SequrDemo.hex_string_to_bytes("01020304050607080901020304050607")!        // Note: Generates random 16 bytes iv per encrypt operation or per message. Currently fixed for this demo
        
        // Encrypts message with QEEP wrapper class
        let ret: Int32 = qeep_wrapper!.encrypt(
            key: &(qk!),
            iv: &message_iv!,
            message: &message_bytes,
            encrypted_message: &encrypted!
        )
        
        // Handle output based on encrypt result
        if (ret == 1){
            // Converts byte array into hex string for printing
            return String(Data(encrypted!).hexEncodedString())
        }else{
            return "Error code: \(ret) occurred in encryption. Consult error.h for more information"
        }
    }
    
    /// 'Decrypt' button press event
    func decrypt_btn_pressed(){
        // Perform the decryption
        decrypted_message_string = decrypt_operation(key: &(qk), iv: &message_iv, encrypted_message: &encrypted_message_bytes, message: &decrypted_message_bytes)
    }
    
    /// Performs decryption using the QEEP algorithm on a passed-in encrypted byte array
    ///
    ///  - Prerequisites:
    ///   - key (QK): Must be retrieved before attempting encryption/decryption
    ///   - iv: Must be setup in order to use during encryption process
    ///  - Parameters:
    ///    - key (QK): the same cryptographic key used to perform the decryption
    ///    - iv: initialization vector to improve randomness
    ///    - encrypted_message: the encrypted message as a result of pressing the encrypt button
    ///    - message: the plaintext message that results from QEEP decrypt
    ///  - Returns: A QEEP decrypted form of the encrypted message or the issue being faced
    func decrypt_operation(key qk:inout [UInt8]?, iv message_iv: inout[UInt8]?, encrypted_message encrypted_message_bytes: inout [UInt8]?, message decrypted_message_bytes: inout [UInt8]?) -> String{
        // Verifies that QK is present before attempting to encrypt/decrypt
        if(qk == nil){
            return "Need to retrieve QK before decrypting"
        }
        if(encrypted_message_bytes == nil){
            return "Need to encrypt before decrypting"
        }
        
        // Decrypts the encrypted message
        // Note: Message iv should be identical to one used during encryption
        let ret: Int32 = qeep_wrapper!.decrypt(
            key: &qk!,
            iv: &message_iv!,
            encrypted_message: &encrypted_message_bytes!,
            message: &decrypted_message_bytes!
        )

        // Handles UI display according to error codes
        if (ret == 1){
            // Outputs the byte array as string to screen
            return String(bytes: decrypted_message_bytes!, encoding: .utf8)!
        }else{
            return "Error code: \(ret) occurred in encryption. Consult error.h for more information"
        }
    }
    
    @Environment(\.presentationMode) var presentationMode: Binding<PresentationMode>
    var btn_back : some View {
        Button(action:{
            self.presentationMode.wrappedValue.dismiss()
        }) {
            HStack{
                Image("back_button_image")
                    .aspectRatio(contentMode: .fit)
                    .foregroundColor(Color.white)
                Text("Back")
            }
        }
    }
    
    var body: some View {
        
        ZStack{
            qeep_demo_BG_colour
                .ignoresSafeArea()
            VStack{
                HStack{
                    Spacer().frame(width: 10)
                    
                    Button(action: get_qk){
                        Text("GET QK")
                            .padding()
                            .background(Color.gray)
                            .foregroundColor(.white)
                    }
                    Spacer()
                }
                Spacer()
                HStack{
                    VStack{
                        TextField("Enter message", text: $user_message)
                            .keyboardType(.default)
                            .padding(.horizontal, 30).padding(.top, 20)
                            .multilineTextAlignment(TextAlignment.center)
                            .foregroundColor(Color.white)
                        
                        Divider()
                            .padding(.horizontal, 30)
                            .background(Color.white)
                    }
                    
                    Button(action: encrypt_btn_pressed){
                        Text("ENCRYPT")
                            .padding()
                            .background(Color.gray)
                            .foregroundColor(.white)
                    }
                }
                Spacer()
                
                HStack{
                    Text("Encrypted Message:")
                    Text(encrypted_message_string)
                }.foregroundColor(Color.white)
                
                Spacer()
                
                Button(action: decrypt_btn_pressed){
                    Text("DECRYPT")
                        .padding()
                        .background(Color.gray)
                        .foregroundColor(.white)
                }
                
                Spacer()
                
                HStack{
                    Text("Decrypted Message: ")
                    Text(decrypted_message_string)
                }.foregroundColor(Color.white)
                Spacer()
               
            }.padding()
        }
        .navigationBarBackButtonHidden(true)
        .navigationBarItems(leading: btn_back)
    }
}

//struct QeepDemo_Previews: PreviewProvider {
//    static var previews: some View {
//        let requests = Requests()
//        QeepDemo().environmentObject(requests)
//    }
//}
